portfolio/src/main/java/com/google/sps/servlets/ServeCommentsServlet.java

```diff
@@ -45,19 +66,52 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String email = request.getParameter("email");
     String content = request.getParameter("content");
-    long timestamp = System.currentTimeMillis();
+    String userTokenId = extractUserTokenId(request);
+
+    try {
+      GoogleIdToken idToken = validateUserIdToken(userTokenId);
+      System.out.println("Token ID is valid.");
+      addCommentToDb(idToken, content);
+      response.sendRedirect("/leave-comment/leave-comment.html");
+    } catch (GeneralSecurityException e) {
+      System.out.println("Token ID was not valid.");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
+    }
+  }
+
+  private static String extractUserTokenId(HttpServletRequest request) {
+    String userTokenId = request.getParameter("userTokenId");
+    if (userTokenId == null) {
+      throw new IllegalArgumentException("User token ID is empty.");
+    }
+    return userTokenId;
+  }
 
+  private void addCommentToDb(GoogleIdToken idToken, String content) {
+    String email = idToken.getPayload().getEmail();
+    long timestamp = System.currentTimeMillis();
     Entity commentEntity = new Entity("Comment");
     commentEntity.setProperty("email", email);
     commentEntity.setProperty("content", content);
     commentEntity.setProperty("timestamp", timestamp);
-
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
+  }
+
+  private GoogleIdToken validateUserIdToken(String idTokenString) throws GeneralSecurityException {
+    GoogleIdToken idToken;
+    try {
+      idToken = verifier.verify(idTokenString);
+    } catch (IOException e) {
+      throw new GeneralSecurityException("User token ID is not valid");
```

Comment:
You can propagate `e` as the cause of this `GeneralSecurityException`

Reply:
I made attempt to propagate it and include it in info. Not sure if it's the best way to handle this exception, but it works.